src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -93,10 +76,38 @@ public int size() {
     }
 
     public InetSocketAddress next(long spinDelay) {
-        ++currentIndex;
-        if (currentIndex == serverAddresses.size()) {
-            currentIndex = 0;
+        if (!connectedSinceNext) {
+            // The current address was generated by the previous call to
+            // next(). That call, did not result in a successful connection.
+            //
+            // Therefore, attempt re-resolution on the current address
+            // before moving on to the next index.
+            InetSocketAddress curAddress = serverAddresses.get(currentIndex);
+            LOG.info("Re-resolving {}", curAddress.getHostString());
+
+            // Re-resolve the address
+            try {
+                List<InetSocketAddress> reresolvedAddresses = resolveAddress(curAddress);
+
+                if (reresolvedAddresses.size() == 1) {
+                    serverAddresses.set(currentIndex, reresolvedAddresses.get(0));
+                } else {
+                    serverAddresses.remove(currentIndex);
```

Comment:
So we don't want to replace the address at `currentIndex`, we just want to add it to the list. Does that not mean that we effectively skip an address if we execute this branch and then execute the `currentIndex++` below?

Reply:
The PR implements what is basically lazy re-resolution.

Let's say there are 3 addresses, all initially resolved:
```0 star.net => 1.1.1.1
1 circle.net => 2.2.2.2
2 square.net => 3.3.3.3```

The very first time `next()` is called, `star.net => 1.1.1.1` is returned. Then, let's say the client is disconnected _and_ the original resolution is now stale:

client disconnected
calls `next()`
receives circle.net => 2.2.2.2
tries, and fails to connect
calls `next()` again
within `next()`, circle.net is re-resolved to 4.4.4.4
index 1 is now set to circle.net => 4.4.4.4
`next()` returns index 2, square.net => 3.3.3.3
tries, and fails to connect
calls `next()` again
within `next()`, square.net is re-resolved to 5.5.5.5
index 2 is now set to square.net => 5.5.5.5
`next()` returns index 0, star.net => 1.1.1.1
tries, and fails to connect
calls `next()` again
within `next()`, star.net is re-resolved to 6.6.6.6
`next()` returns index 1, circle.net => 4.4.4.4
Tries, and succeeds in connecting

In the branch you highlight above, where re-resolved addresses are appended and the original address removed we'd have:
 
client disconnected
calls `next()`
receives circle.net => 2.2.2.2
tries, and fails to connect
calls `next()` again
within `next()`, circle.net is re-resolved to 4.4.4.4 and 4.4.4.3

So the list is now:
```
0 star.net => 1.1.1.1
1 square.net => 3.3.3.3
2 circle.net => 4.4.4.4
3 circle.net => 4.4.43
```

So yes, it would skip ever trying square.net => 3.3.3.3. I'll add a -- to the index in that branch.

Reply:
Would an alternative be something like:
```java
                serverAddresses.set(currentIndex, reresolvedAddresses.get(0));
                
                resolvedAddresses.remove(0);
                for (InetSocketAddress resolvedAddress : reresolvedAddresses) {
                     if (!serverAddresses.contains(resolvedAddress)) {
                          serverAddresses.add(resolvedAddress);
                      }
                }
```
which effectively combines the two branches. Would this be better than index manipulations?

Reply:
Yes, good catch. Modifying...